### PR TITLE
Endleaf rejiggle, .fsm syntax for endids, refactor endid get/set api, various related bugfixes

### PR DESCRIPTION
**Pull request by katef**
I did quite a few things here, all focused around the endleaf callbacks,
and by extension the endid mechanism. I fixed a few small bugs, and got distracted trying to tidy up a bit.

Anyway here's what we get:

* .fsm syntax support for endids
* we now have tests for .fsm syntax (gasp)
* Add missing printing for endids for various formats (notably dot and .fsm)
* Various work on ways to simplify the endids API, especially:
  * Removed &nwritten for fsm_endid_get(), we don't need partial gets
  * Removed the tri-state enums for fsm_endid_get/set()
  * We don't need to distinguish a state absent endids from a present but empty set of endids
* Add missing calls to endleaf() callbacks for various formats
* Reintroduced fsm_exec handling in re(1), and reworked related argv[] handling. This is neccessary bugfixes only, in particular I want to address #478 and #477 separately
* Bugfixes around the api format codegen, that was quite out of date.

The .fsm syntax for endids looks like this:
```
; ./build/bin/re -r pcre -bpl fsm 'ab?c' 'abc*'
0  ->  1 "a"; # e.g. "a"
1  ->  2 "b"; # e.g. "ab"
1  ->  3 "c"; # e.g. "ac"
2  ->  4 "c"; # e.g. "abc"
4  ->  5 "c"; # e.g. "abcc"
5  ->  5 "c"; # e.g. "abcc"

start: 0;
end:   2 = [0], 3 = [0], 4 = [0, 1], 5 = [0];
```
and renders out like:
image

### CODE REVIEW

**Review by silentbicycle — CHANGES_REQUESTED** [review]
Do you have a better suggestion for how to pass a handle to the caller besides `void *opaque` that isn't significantly more complicated in a multi-threaded context?

File: src/libfsm/print/c.c [hunk 1]
```diff
@@ -425,29 +431,26 @@ fsm_print_c_complete(FILE *f, const struct ir *ir,
...
 			fprintf(f, "{\n");
-			fprintf(f, "\t(void)opaque;\n");
 			if (ir->n > 0) {
 				fprintf(f, "\tint c;\n");
 				fprintf(f, "\n");
 			}
 			break;
 
 		case FSM_IO_STR:
-			fprintf(f, "(const char *s, void *opaque)\n");
```
[the remark this answers is not in the document]
Comment (katef) [reply src/libfsm/print/c.c:438]: You could still use a `void *opaque` if you like! But my point is with `.fragment = true` output, that's the responsibility of the API-user, it's not something that belongs to libfsm's generated code. That's why I'm removing it here.

File: include/fsm/fsm.h [hunk 2]
```diff
@@ -222,29 +216,31 @@ fsm_setendid(struct fsm *fsm, fsm_end_id_t id);
...
-	FSM_GETENDIDS_FOUND,
-	FSM_GETENDIDS_ERROR_INSUFFICIENT_SPACE = -1
-};
-enum fsm_getendids_res
-fsm_getendids(const struct fsm *fsm, fsm_state_t end_state,
-    size_t id_buf_count, fsm_end_id_t *id_buf,
-    size_t *ids_written);
+int
+fsm_endid_get(const struct fsm *fsm, fsm_state_t end_state,
+    size_t id_buf_count, fsm_end_id_t *id_buf);
```
[the remark this answers is not in the document]
Comment (katef) [reply include/fsm/fsm.h:239]: Right, and I dealt with those situations.

[the remark this answers is not in the document]
Comment (silentbicycle) [reply include/fsm/fsm.h:239]: Yes, just providing context.

File: src/libfsm/print/c.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (silentbicycle) [reply src/libfsm/print/c.c:438]: Okay, that works for me.

**Review by silentbicycle — APPROVED** [review]

**Pull request merged**